src/main/java/io/stargate/sgv2/jsonapi/service/operation/model/impl/ReadAndUpdateOperation.java

```diff
@@ -136,29 +141,47 @@ public Uni<Supplier<CommandResult>> execute(
 
   private Uni<UpdatedDocument> processUpdate(
       DataApiRequestInfo dataApiRequestInfo,
-      ReadDocument document,
+      ReadDocument readDocument,
       QueryExecutor queryExecutor,
       AtomicInteger modifiedCount) {
     return Uni.createFrom()
-        .item(document)
-
-        // perform update operation and save only if data is modified.
+        .item(readDocument)
         .flatMap(
-            readDocument -> {
-              // if there is no document return null item
+            document -> {
+              // if there is no document: return null item
               if (readDocument == null) {
                 return Uni.createFrom().nullItem();
               }
-
               // upsert if we have no transaction if before
               boolean upsert = readDocument.txnId() == null;
-              JsonNode originalDocument = upsert ? null : readDocument.document();
-
-              // apply document updates
-              // if no changes return null item
-              DocumentUpdater.DocumentUpdaterResponse documentUpdaterResponse =
```

Comment:
Here is a good overlook for the logic.
We do two levels of update now.
First level, update anything else, but do not re-vectorize yet.
Second level, will do vectorization on demand and update the $vector accordingly.